### ISSUE TITLE
visibilityjs causes the app to crash if running in node

### ISSUE THREAD

**Issue by ingara** [GetStream/stream-chat-react#196]
I'm using NextJS with SSR, and because code is shared between client and server, `visibilityjs` causes issues.

If I have any module that imports something from `stream-chat-react` it fails server side because of `visibilityjs`. This lib initializes itself with an IIFE that takes `window` as an argument, and as such it will cause a runtime error if any code that causes this module to be evaluated is used.

I can work around this in some cases by using NextJS's dynamic imports, which can be configured to always render a fallback on the server side, but this is not a feasible workaround in all cases.

Would you consider swapping `visibilityjs` for some other lib that behaves more gracefully? Alternatively, maybe something can be done about how and when it is loaded?

**Comment by vishalnarkhede** [comment]
Hey @ingara , I will look into it and get back to you!

**Comment by hansy** [comment]
I'm having the same issue, but also with the `blueimp-load-image` module. In this case, `document is not defined`. I'm using static site generation with Gatsby.js, but the build phase errors out here.

**Comment by vishalnarkhede** [comment]
Hey @hansy @ingara you can disable the SSR for chat components. e.g,
```
import NoSSR from 'react-no-ssr'
import { StreamChat } from 'stream-chat';
export default function Index() {
  return (
    <div>
      <p>Hello this is Chat app</p>
      <NoSSR onSSR={<div>Loading ...</div>}>
        <ChatAppComponent />
      </NoSSR>
    </div>
  );
}
function ChatAppComponent() {
    const {
        Chat,
        ChannelList,
    } = require('stream-chat-react')
    const chatClient = new StreamChat('qk4nn7rpcn75');
    chatClient.setUser(
      {
        id: 'example-user',
      },
      'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyX2lkIjoiZXhhbXBsZS11c2VyIn0.HlC0dMKL43y3K_XbfvQS_Yc3V314HU4Z7LrBLil777g',
    );
    return (
      <div>
        <p>This is the about page</p>
        <Chat client={chatClient} theme={`messaging`}>
            <ChannelList
            />
      </Chat>
      </div>
    );
  }
```

**Comment by ingara** [comment]
@vishalnarkhede Yes, that is what I'm doing as a workaround at the moment. It would be great not having to do that though, as it feels hacky and is confusing to newcomers to the library.

**Comment by hansy** [comment]
Agree with @ingara. I've been doing something equally hacky with dynamic imports a la:

```
let Chat;

class Test extends React.Component {
  constructor(props) {
    super(props)
    this.state = { loaded: false }
  }

  componentDidMount() {
    import('../components/chat.js')
      .then((res) => {
        Chat = res.default
        this.setState({ loaded: true })
      })
  }

  render() {
    this.state.loaded ? return <Chat /> : return <p>Not loaded</p>
  }
}

export default Test;
```

**Comment by vishalnarkhede** [comment]
@ingara @hansy I can understand why it feels hacky. But rendering chat on serverside itself doesn't make much sense. Even if we make changes to somehow handle the issue of visibilityjs or document being unavailable - you will still need to establish the websocket connection and then query channels or channel on client side (which is pretty much the gist of chat application).

What do you think?

_This is just my personal opinion ... open for discussion :)_

**Comment by ingara** [comment]
Oops, misclicked the close button, sorry.

I totally agree that rendering the chat server-side doesn't make sense, but having this limitation (not being able to reference any module that directly or transitively imports something from `stream-chat-react`) does make structuring the code harder.

For instance, I need to display a badge with the count of unread chat messages in our menu. The menu itself has to be rendered server-side, but I can defer getting the unread count until client-side. If I was able to have a file with the import in it without it crashing I could create a simple function (or hook) that conditionally returns 0 server-side or the actual number client-side.
Since this is not possible I have to change my React component structure so that I can use either `react-no-ssr` as you mention, or `next/dynamic`. I haven't found a way to use regular dynamic imports in my use-case yet.

**Comment by vishalnarkhede** [comment]
@ingara fair enough. 

I will keep this issue open for now. I will try (within next couple of weeks) to look into dependencies that are causing this issue and see if we can offer some solution OOTB.

**Comment by fandy** [comment]
Got stuck on the same issue as @hansy with Next.js SSG. Wrapping the Chat component with NoSsr results in the same issue, complaining about `document is not defined` in the `blueimp-load-image`module.

**Comment by jaapbakker88** [comment]
@fandy @ingara @hansy we've just released a new version that includes a fix for this, it's on version `2.1.1`

#466

**Comment by thehappycoder** [comment]
With the latest version of stream-chat-react, there are SSR issues again. Managed to workaround the issue by dynamically importing both stream and react stream packages (both are SSR unfriendly atm) and passing them to all children that need them. NoSSR package didn’t help because the problem happens as soon as a library is imported, not when components are rendered.
